rust/arrow/src/array/array_binary.rs

```diff
@@ -258,6 +258,8 @@ where
             }
         }
 
+        // calculate actual data_len, which may be different from the iterator's upper bound
+        let data_len = offsets.len() - 1;
```

Comment:
I wonder if there is a problem of writing off the end of `null_buf` which [was allocated](https://github.com/apache/arrow/pull/9588/files#diff-0b271702d55fe980ca7906ba50e561f00e610ce3de70ea1b0782e1d6be6f60c3R252) to be only size of `data_len`?

Reply:
writing beyond the allocated capacity would only happen if the iterator input returns more items than the declared upper bound - this would be against the `Iterator` spec and I think that would be unlikely ;   
this PR addresses the opposite case where an iterator returns less items than its declared upper bound (which is still within the `Iterator` spec);

Reply:
I think that @alamb is right (my comment is actually wrong). @yordan-pavlov , the Iterator spec recommends, does not require, that the iterator reports a correct length. Consumer that lead to undefined behavior from an incorrect `size_hint` are the causers of said undefined behavior.

The only case where consumers can trust the iterators' length is when the interator implement `unsafe trait TrustedLen`. Unfortunately, `TrustedLen` is still in unstable. For that reason, we have been exposing `unsafe Buffer::from_trusted_len_iter` and the like for those cases.

The issue here is that to correctly declare a FromIterator in this case, we must use `BooleanBufferBuilder` for the null bitmap, which is significantly slower than a `MutableBuffer`. This is why I added that `panic!` there, even though it is unsound. This PR is just surfacing this problem where the current implementation is `unsound` (because the iterator can be larger than `upper_bound`).

Since we have been seeing people prefer the `FromIter` over the `Builder` APIs, my suggestion is that we use `BooleanBufferBuilder`, like we are doing for `PrimitiveArray`, and offer (in a future PR), a `from_trusted_len_iter` for string arrays.

Reply:
you are right @jorgecarleitao, it is up to the iterator reader to ensure incorrectly declared bounds do not result in incorrect behavior and, sadly the current implementation could write out of allocated memory in the case of an iterator that produces more items than the declared upper bound; 
but although I agree this should be fixed, I think it falls outside of the scope of this PR.

Reply:
After some more study I agree

Reply:
https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.size_hint  says

> It is not enforced that an iterator implementation yields the declared number of elements. A buggy iterator may yield less than the lower bound or more than the upper bound of elements.

> size_hint() is primarily intended to be used for optimizations such as reserving space for the elements of the iterator, but must not be trusted to e.g., omit bounds checks in unsafe code. An incorrect implementation of size_hint() should not lead to memory safety violations.

I figured i would paste that into this PR as I had looked it up to satisfy my own curiosity